### PR TITLE
feat(auth): add setUpTOTP API

### PR DESCRIPTION
**Pull request by israx**
#### Description of changes

`setUpTOTP` API is used to kickoff the TOTP  process. It returns a `sharedSecret` that can be input directly on authenticator apps to generate an OTP code.

It also returns a `getSetupUri` util function, which generates a link that can be used to create a QR code. This QR code can be read by authenticator apps and start generating OTP codes on demand. 

**TODO list:**
- get the decoded `idToken` and grab username to be used by the `getSetupUri` util method.
- get the `accessToken` from the AuthToken provider.

#### Description of how you validated changes

In my local machine, I used the `CognitoUserPoolTokenManager.setTokens` method to set tokens on a successful call of `signIn`. Then I used the `CognitoUserPoolTokenManager.loadTokens` method inside `setUpTOTP` API to grab the `accessToken` and pass it to the client.

Then I created a sample App to test the functionality, and was able to associate cognito with an authenticator App.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/auth/src/providers/cognito/apis/setUpTOTP.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { AuthError } from '../../../errors/AuthError';
+import { TOTPSetupDetails } from '../../../types/models';
+import { SETUP_TOTP_EXCEPTION,  AssociateSoftwareTokenException } from '../types/errors';
+import { associateSoftwareTokenClient } from '../utils/clients/AssociateSoftwareTokenClient';
+import { getTOTPSetupDetails } from '../utils/signInHelpers';
+/**
```
Comment (jimblanc) [inline packages/auth/src/providers/cognito/apis/setUpTOTP.ts:9]:
```suggestion

/**
```

File: packages/auth/src/providers/cognito/apis/setUpTOTP.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * Sets up TOTP for the user.
+ * @throws  service error thrown when setting up TOTP: {@link AssociateSoftwareTokenException}.
```
Comment (jimblanc) [inline packages/auth/src/providers/cognito/apis/setUpTOTP.ts:11]: Nit: Technically I think the format should look like this according to TSDocs, but if it looks good in your IDE should be fine.

```suggestion
 *
 * @throws {@link AssociateSoftwareTokenException}
 * Thrown is a service occurs while setting up TOTP.
```

**Review by jimblanc — DISMISSED** [review]
LGTM, couple nits below.

File: packages/auth/__tests__/providers/cognito/setUpTOTP.test.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	});
+
+	test('setUpTOTP API should call the UserPoolClient and should return a TOTPSetupDetails', async () => {
+		const result = await setUpTOTP();
+		expect(result.sharedSecret).toEqual(secretCode);
+		expect(result.getSetupUri('appName', 'amplify')).toBeInstanceOf(URL);
+	});
+});
+
+describe('setUpTOTP API error path cases:', () => {
```
Comment (kvramyasri7) [inline packages/auth/__tests__/providers/cognito/setUpTOTP.test.ts:37]: Can we include a test case where we do not have `SecretCode` as well?

**Pull request merged**